File: resources/js/ode/julia/solvers/esdirk.js
/* ==========================================================================
   ode_julia / solvers / esdirk

   Singly diagonally implicit Runge-Kutta methods with an explicit first stage:
   TRBDF2 and KenCarp4.

   Each stage solves, for the increment zᵢ = h·f at that stage,

       zᵢ = h·f(t + cᵢh,  uprev + Σⱼ<ᵢ aᵢⱼ·zⱼ + γ·zᵢ)

   by the simplified Newton iteration in core/newton.js. "Singly" means every
   stage shares the same γ, so one W = I − γh·J serves the whole step and, if
   the step size does not move much, several steps after it. That is the
   economy these methods are built around, and it is why they beat a Rosenbrock
   method of the same order on a long smooth run.

   Both are stiffly accurate: the last row of `a` is `b`, so the solution is
   the last stage value, and f at the end of the step is z_s/h -- free, and
   reused as the first stage of the next step.

   Where they sit. TRBDF2 is second order, which sounds modest and is not: it
   is L-stable, it is unusually forgiving on problems where Newton struggles,
   and at loose tolerances it is often the cheapest thing here. KenCarp4 is
   fourth order and is the one to reach for at tolerances below about 1e-6.
   ========================================================================== */

import { TRBDF2Tableau, KenCarp4Tableau } from './esdirk-tableaus.js';
import { DIRK, Convergence } from '../core/newton.js';

class ESDIRKCache {
  constructor(n, tab, integ, opts) {
    this.tab = tab;
    this.n = n;
    this.order = tab.order;
    this.errorOrder = tab.errorOrder;
    // Stiffly accurate: f at the end of the step is the last stage's z over h.
    this.hasFsalLast = true;
    this.dtpropose = null;

    this.z = Array.from({ length: tab.stages }, () => new Float64Array(n));
    this.est = new Float64Array(n);
    this.smoothEst = opts.smoothEst !== false;
    // How many consecutive steps have reused the same W. Julia rebuilds when
    // the Newton iteration stops converging quickly, not on a fixed schedule.
    this.newW = true;
  }

  init(integ) {
    integ.newton.method = DIRK;
    integ.newton.gamma = this.tab.gamma;
  }

  restart() { this.newW = true; }

  step(integ) {
    const { tab, n, z } = this;
    const dt = integ.dt;
    const { uprev, u, t } = integ;
    const newton = integ.newton;
    const gammaDt = tab.gamma * dt;
    const s = tab.stages;

    // One W for the whole step, and usually for several steps after it.
    //
    // Two different questions, and they were confused here at first: W must be
    // rebuilt whenever γh moves, and `form` does that from the stored Jacobian
    // without being asked. A new JACOBIAN is a much larger expense and is
    // justified only by the Newton iteration labouring, or by something having
    // invalidated it. Forcing one on every change of step size -- which is
    // almost every step -- threw away the reuse entirely: FBDF re-differenced
    // on three steps in four.
    const freshJac = integ.W.jacStale || !integ.W.haveFactor || !newton.fastConvergence;
    const gammaMoved = Math.abs(integ.W.gammaDt - gammaDt) > 0.2 * Math.abs(gammaDt);
    if (!integ.formW(gammaDt, false, freshJac)) return false;
    this.newW = freshJac || gammaMoved;

    // Stage 1 is explicit: z₁ = h·f(t, uprev), which the integrator has.
    for (let i = 0; i < n; i++) z[0][i] = dt * integ.fsalfirst[i];

    for (let stage = 1; stage < s; stage++) {
      const a = tab.a[stage];
      const alpha = tab.alpha[stage];
      for (let i = 0; i < n; i++) {
        let v = uprev[i];
        for (let j = 0; j < stage; j++) v += a[j] * z[j][i];
        newton.tmp[i] = v;
      }
      // Seed the iterate from the stages already solved. Shampine's choice;
      // it is worth one Newton iteration a stage and costs a few multiplies.
      for (let i = 0; i < n; i++) {
        let v = 0;
        for (let j = 0; j < stage; j++) v += alpha[j] * z[j][i];
        newton.z[i] = v;
      }
      newton.c = tab.c[stage];
      newton.gamma = tab.gamma;
      const status = newton.solve(
        integ.f, integ.W, t, dt, uprev, { reltol: integ.reltol, abstol: integ.abstolFixed },
        this.newW && stage === 1, true,
      );
      integ.stats.nnonliniter += newton.iter;
      if (status !== Convergence) {
        integ.stats.nnonlinconvfail++;
        // A Newton that will not converge is a statement about W, not about
        // accuracy: the next attempt gets a fresh Jacobian and a shorter step.
        integ.W.markStale();
        return false;
      }
      z[stage].set(newton.z);
      this.newW = false;   // later stages of this step reuse the same W
    }

    // Stiffly accurate, so the solution is the last stage value.
    for (let i = 0; i < n; i++) {
      let v = uprev[i];
      for (let j = 0; j < s; j++) v += tab.b[j] * z[j][i];
      u[i] = v;
    }
    // ... and f at the end of the step came with it.
    const invdt = 1 / dt;
    for (let i = 0; i < n; i++) integ.fsallast[i] = z[s - 1][i] * invdt;

    const est = this.est;
    for (let i = 0; i < n; i++) {
      let v = 0;
      for (let j = 0; j < s; j++) {
        const bt = tab.btilde[j];
        if (bt !== 0) v += bt * z[j][i];
      }
      est[i] = v;
    }
    // Smoothing the estimate through W, as Shampine proposed and Julia does by
    // default. The raw embedded difference is the error of an explicit
    // combination and is badly pessimistic on a very stiff problem -- it sees
    // the fast modes the method is deliberately damping. Passing it through
    // (I − γhJ)⁻¹ damps them in the estimate the same way.
    if (this.smoothEst) integ.solveW(est);
    integ.EEst = integ.errorNorm(est);
    return true;
  }
}

/** TRBDF2: 2nd order, L-stable, 3 stages, trapezoid then BDF2. */
export function TRBDF2(options = {}) {
  return {
    name: 'TRBDF2',
    order: TRBDF2Tableau.order,
    build: (n, integ, opts) => new ESDIRKCache(n, TRBDF2Tableau, integ, { ...options, ...opts }),
    controller: { qmax: 10, qsteadyMin: 1, qsteadyMax: 1.2, ...(options.controller || {}) },
  };
}

/** KenCarp4: 4th order, L-stable, 6 stages, the implicit half of ARK4(3)6L[2]SA. */
export function KenCarp4(options = {}) {
  return {
    name: 'KenCarp4',
    order: KenCarp4Tableau.order,
    build: (n, integ, opts) => new ESDIRKCache(n, KenCarp4Tableau, integ, { ...options, ...opts }),
    controller: { qmax: 10, qsteadyMin: 1, qsteadyMax: 1.2, ...(options.controller || {}) },
  };
}

/** The generic builder, for another ESDIRK tableau of the same shape. */
export function esdirkAlgorithm(tab, options = {}) {
  return {
    name: tab.name,
    order: tab.order,
    build: (n, integ, opts) => new ESDIRKCache(n, tab, integ, { ...options, ...opts }),
    controller: { ...(options.controller || {}) },
  };
}

File: kompartment/python/kompartment/engine/solvers/julia/methods/esdirk.py
"""Singly diagonally implicit Runge-Kutta methods with an explicit first
stage: TRBDF2 and KenCarp4 (``solvers/esdirk.js``).

Each stage solves z_i = h f(t + c_i h, uprev + sum_{j<i} a_ij z_j + gamma z_i)
by the shared Newton iteration against one W = I - gamma h J. Both methods
are stiffly accurate: the solution is the last stage value and f at the end
of the step is z_s/h.
"""

from __future__ import annotations

from typing import Any, Dict, List

import numpy as np

from ..integrator import algorithm
from ..newton import CONVERGENCE, DIRK
from .tableaus import KENCARP4, TRBDF2 as TRBDF2_TABLEAU


class ESDIRKCache:
    def __init__(self, n: int, tab: Any, integ: Any, opts: Dict[str, Any]) -> None:
        self.tab = tab
        self.n = n
        self.order = tab.order
        self.error_order = tab.error_order
        self.has_fsal_last = True
        self.dtpropose = None
        self.z: List[np.ndarray] = [np.zeros(n) for _ in range(tab.stages)]
        self.smooth_est = opts.get('smooth_est') is not False
        self.new_w = True

    def init(self, integ: Any) -> None:
        integ.newton.method = DIRK
        integ.newton.gamma = self.tab.gamma

    def restart(self, integ: Any = None) -> None:
        self.new_w = True

    def step(self, integ: Any) -> bool:
        tab = self.tab
        z = self.z
        dt = integ.dt
        uprev = integ.uprev
        t = integ.t
        newton = integ.newton
        gamma_dt = tab.gamma * dt
        s = tab.stages

        fresh_jac = integ.W.jac_stale or not integ.W.have_factor or not newton.fast_convergence
        gamma_moved = abs(integ.W.gamma_dt - gamma_dt) > 0.2 * abs(gamma_dt)
        if not integ.form_w(gamma_dt, False, fresh_jac):
            return False
        self.new_w = fresh_jac or gamma_moved

        z[0] = dt * integ.fsalfirst

        for stage in range(1, s):
            a = tab.a[stage]
            alpha = tab.alpha[stage]
            v = uprev.copy()
            for j in range(stage):
                v += a[j] * z[j]
            newton.tmp[:] = v
            seed = np.zeros(self.n)
            for j in range(stage):
                seed += alpha[j] * z[j]
            newton.z[:] = seed
            newton.c = tab.c[stage]
            newton.gamma = tab.gamma
            status = newton.solve(integ.f, integ.W, t, dt, uprev,
                                  {'reltol': integ.reltol, 'abstol': integ.abstol_fixed},
                                  self.new_w and stage == 1, True)
            integ.stats['nnonliniter'] += newton.iter
            if status != CONVERGENCE:
                integ.stats['nnonlinconvfail'] += 1
                integ.W.mark_stale()
                return False
            z[stage] = newton.z.copy()
            self.new_w = False

        u = uprev.copy()
        for j in range(s):
            u += tab.b[j] * z[j]
        integ.u[:] = u
        invdt = 1 / dt
        integ.fsallast[:] = z[s - 1] * invdt

        est = np.zeros(self.n)
        for j in range(s):
            bt = tab.btilde[j]
            if bt != 0:
                est += bt * z[j]
        if self.smooth_est:
            est = integ.solve_w(est)
        integ.eest = integ.error_norm(est)
        return True


def _controller(options: Dict[str, Any]) -> Dict[str, Any]:
    c = {'qmax': 10, 'qsteady_min': 1, 'qsteady_max': 1.2}
    c.update(options.get('controller') or {})
    return c


def _build(tab: Any, options: Dict[str, Any]) -> Any:
    def build(n: int, integ: Any, opts: Dict[str, Any]) -> ESDIRKCache:
        merged = dict(options)
        merged.update(opts)
        return ESDIRKCache(n, tab, integ, merged)
    return build


def TRBDF2(**options: Any) -> Any:
    return algorithm('TRBDF2', TRBDF2_TABLEAU.order, _build(TRBDF2_TABLEAU, options), _controller(options))


def KenCarp4(**options: Any) -> Any:
    return algorithm('KenCarp4', KENCARP4.order, _build(KENCARP4, options), _controller(options))


def esdirk_algorithm(tab: Any, **options: Any) -> Any:
    return algorithm(tab.name, tab.order, _build(tab, options), dict(options.get('controller') or {}))
